schemachange/cli.py

```diff
@@ -140,72 +186,313 @@ def redact(self, context: str) -> str:
+      )
+      # if there is an error passed with the reponse include that
+      if 'error_description' in resJsonDict.keys():
+        errormessage += _err_oauth_tk_err.format( desc = resJsonDict['error_description']) 
+      raise KeyError( errormessage )
+
+  def authenticate(self):
+    # Password authentication is the default
+    snowflake_password = None
+    if os.getenv("SNOWFLAKE_PASSWORD") is not None and os.getenv("SNOWFLAKE_PASSWORD"):
+      snowflake_password = os.getenv("SNOWFLAKE_PASSWORD")
+    
+    # Check legacy/deprecated env variable
+    if os.getenv("SNOWSQL_PWD") is not None and os.getenv("SNOWSQL_PWD"):  
+      if snowflake_password:
+        warnings.warn(_warn_password_dup, DeprecationWarning)
+      else:
+        warnings.warn(_warn_password, DeprecationWarning)
+        snowflake_password = os.getenv("SNOWSQL_PWD")
+
+    if snowflake_password:
+      if self.verbose:
+        print(_log_auth_type %  'password' )
+      self.conArgs['password'] = snowflake_password
+      self.conArgs['authenticator'] = 'snowflake'
+
+    # If no password, try private key authentication 
+    elif os.getenv("SNOWFLAKE_PRIVATE_KEY_PATH", ''):
+      if self.verbose:
+        print( _log_auth_type %  'private key')
+
+      private_key_password = os.getenv("SNOWFLAKE_PRIVATE_KEY_PASSPHRASE", '')
+      if private_key_password:
+        private_key_password = private_key_password.encode()
+      else:
+        private_key_password = None
+        if self.verbose:
+          print(_log_pk_enc)
+      with open(os.environ["SNOWFLAKE_PRIVATE_KEY_PATH"], "rb") as key:
+        p_key= serialization.load_pem_private_key(
+            key.read(),
+            password = private_key_password,
+            backend = default_backend()
+        )
+
+      pkb = p_key.private_bytes(
+          encoding = serialization.Encoding.DER,
+          format = serialization.PrivateFormat.PKCS8,
+          encryption_algorithm = serialization.NoEncryption())
+
+      self.conArgs['private_key'] = pkb 
+      self.conArgs['authenticator'] = 'snowflake'
+    
+    elif os.getenv("SNOWFLAKE_AUTHENTICATOR") == 'oauth' and os.getenv("SNOWFLAKE_AUTHENTICATOR"):      
+      oauth_token = self.get_oauth_token()
+      
+      if self.verbose:
+        print( _log_auth_type %  'Oauth Access Token')
+      self.conArgs['token'] = oauth_token
+      self.conArgs['authenticator'] = 'oauth'
+    
+    elif os.getenv("SNOWFLAKE_AUTHENTICATOR") == 'externalbrowser' and os.getenv("SNOWFLAKE_AUTHENTICATOR"):
+      self.conArgs['authenticator'] = 'externalbrowser'
+      if self.verbose:
+        print(_log_auth_type %  'External Browser')
+        
+    elif (os.getenv("SNOWFLAKE_AUTHENTICATOR").lower()[-9:] == '.okta.com' \
+      or os.getenv("SNOWFLAKE_AUTHENTICATOR").lower()[-16:] == '.oktapreview.com'  ) \
+      and os.getenv("SNOWFLAKE_AUTHENTICATOR"):
+      okta = os.getenv("SNOWFLAKE_AUTHENTICATOR")
+      self.conArgs['authenticator'] = okta
```

Comment:
@MACKAT05 Could we check for 'https://' instead of okta string. That way we can avoid limiting it two endpoint patterns but allow the end user to use the Okta endpoint set up to authenticate through SSO. I have a PR #140 for 3.4.3 enabling SSO via Okta.

Reply:
I think that is possible; i was not sure if okta as implemented by the connector was a generic standard. and looking into it further it does look to support that.

so `'os.getenv("SNOWFLAKE_AUTHENTICATOR").lower()[:4]=='http'`  as a single string check and include in the readme that the url must not be malformed?   
I looked for a simple python check if string is an url but I did not find one that was straightforward and did not create additional dependencies.  
would the  shorter string check to include those whom may inadvisably be testing on non https endpoints.

Reply:
would look for `https://` instead of just `http`. That single string check should ensure that it is secure URL. If the URL is invalid, then python connector for snowflake should return an error that the url endpoint is invalid. It is better to enforce the expectation of https than open http options.

Reply:
Yes. Mentioning it in the readme file that URL should not be malformed otherwise, the authentication will not work.